Support for RISC-V architecture

Patch for RISC-V (64 bit only) support:

http://oirase.annexia.org/tmp/numpy-riscv64.patch

> If the patch has only been compile-tested, then I think it should not be applied.

I backported the patch to the Debian package, so on version 1.13.3. It built fine on riscv64 and the testsuite is almost fine too:

Python 2.7: 
```
Ran 6601 tests in 1307.277s

FAILED (KNOWNFAIL=12, SKIP=24, failures=2)
```

Python 3.6:
```
Ran 6610 tests in 1493.899s

FAILED (KNOWNFAIL=13, SKIP=15, failures=2)
```

In both cases the failures are the same:

```
======================================================================
FAIL: test_float (test_numeric.TestBoolCmp)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/debian/tmp/usr/lib/python2.7/dist-packages/numpy/core/tests/test_numeric.py", line 396, in test_float
    assert_array_equal(np.signbit(self.signf[i:]), self.ef[i:])
  File "/<<PKGBUILDDIR>>/debian/tmp/usr/lib/python2.7/dist-packages/numpy/testing/utils.py", line 854, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')
  File "/<<PKGBUILDDIR>>/debian/tmp/usr/lib/python2.7/dist-packages/numpy/testing/utils.py", line 778, in assert_array_compare
    raise AssertionError(msg)
AssertionError:
Arrays are not equal

(mismatch 4.296875%)
 x: array([False, False, False, False, False, False, False, False,  True,
       False, False, False, False, False, False, False, False,  True,
       False, False, False, False, False, False,  True,  True, False,...
 y: array([False, False, False, False, False, False, False, False,  True,
       False, False, False, False, False, False, False, False,  True,
       False, False, False, False, False, False,  True,  True, False,...

======================================================================
FAIL: test_scripts.test_f2py
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/<<PKGBUILDDIR>>/debian/tmp/usr/lib/python2.7/dist-packages/numpy/testing/decorators.py", line 147, in skipper_func
    return f(*args, **kwargs)
  File "/<<PKGBUILDDIR>>/debian/tmp/usr/lib/python2.7/dist-packages/numpy/tests/test_scripts.py", line 93, in test_f2py
    assert_(success, msg)
  File "/<<PKGBUILDDIR>>/debian/tmp/usr/lib/python2.7/dist-packages/numpy/testing/utils.py", line 92, in assert_
    raise AssertionError(smsg)
AssertionError: Warning: neither f2py nor f2py2 nor f2py2.7 found in path

----------------------------------------------------------------------
```
The first failure is specific to RISC-V and probably needs some investigation. The second failure is due to f2py missing during the build, and also happens when building the Debian package on other architectures.

I guess the per-arch flags are correct, otherwise I would have expected many more failures.

--- a/doc/release/1.15.0-notes.rst
+++ b/doc/release/1.15.0-notes.rst
@@ -82,6 +82,10 @@ Creating a full iOS-compatible NumPy package requires building for the 5
 architectures supported by iOS (i386, x86_64, armv7, armv7s and arm64), and
 combining these 5 compiled builds products into a single "fat" binary.
 
+Build system
+------------
+Added experimental support for the 64-bit RISC-V architecture.
+
 Improvements
 ============
 

--- a/numpy/core/include/numpy/npy_cpu.h
+++ b/numpy/core/include/numpy/npy_cpu.h
@@ -17,6 +17,7 @@
  *              NPY_CPU_SH_BE
  *              NPY_CPU_ARCEL
  *              NPY_CPU_ARCEB
+ *              NPY_CPU_RISCV64
  */
 #ifndef _NPY_CPUARCH_H_
 #define _NPY_CPUARCH_H_
@@ -82,6 +83,8 @@
     #define NPY_CPU_ARCEL
 #elif defined(__arc__) && defined(__BIG_ENDIAN__)
     #define NPY_CPU_ARCEB
+#elif defined(__riscv) && defined(__riscv_xlen) && __riscv_xlen == 64
+    #define NPY_CPU_RISCV64
 #else
     #error Unknown CPU, please report this to numpy maintainers with \
     information about your platform (OS, CPU and compiler)

--- a/numpy/core/include/numpy/npy_endian.h
+++ b/numpy/core/include/numpy/npy_endian.h
@@ -46,7 +46,7 @@
             || defined(NPY_CPU_SH_LE)   \
             || defined(NPY_CPU_MIPSEL)  \
             || defined(NPY_CPU_PPC64LE) \
-            || defined(NPY_CPU_ARCEL)	\
+            || defined(NPY_CPU_ARCEL)   \
             || defined(NPY_CPU_RISCV64)
         #define NPY_BYTE_ORDER NPY_LITTLE_ENDIAN
     #elif defined(NPY_CPU_PPC)          \

--- a/numpy/core/include/numpy/npy_endian.h
+++ b/numpy/core/include/numpy/npy_endian.h
@@ -46,7 +46,8 @@
             || defined(NPY_CPU_SH_LE)   \
             || defined(NPY_CPU_MIPSEL)  \
             || defined(NPY_CPU_PPC64LE) \
-            || defined(NPY_CPU_ARCEL)
+            || defined(NPY_CPU_ARCEL)	\
+            || defined(NPY_CPU_RISCV64)
         #define NPY_BYTE_ORDER NPY_LITTLE_ENDIAN
     #elif defined(NPY_CPU_PPC)          \
             || defined(NPY_CPU_SPARC)   \

